inference-engine/tests/functional/inference_engine/ov_async_infer_queue.cpp

```diff
@@ -0,0 +1,295 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <gtest/gtest.h>
+#include <gmock/gmock-matchers.h>
+
+#include <map>
+
+#include <openvino/core/function.hpp>
+#include <openvino/core/shape.hpp>
+#include <openvino/core/type/element_type.hpp>
+
+#include <openvino/op/add.hpp>
+#include <openvino/op/multiply.hpp>
+#include <openvino/op/parameter.hpp>
+
+#include <openvino/runtime/core.hpp>
+#include <openvino/runtime/infer_request.hpp>
+#include <openvino/runtime/executable_network.hpp>
+#include <openvino/runtime/tensor.hpp>
+#include <openvino/runtime/async_infer_queue.hpp>
+
+using namespace ::testing;
+using namespace std;
+
+std::shared_ptr<ov::Function> create_add_model(ov::Shape shape) {
+    auto arg0 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto arg1 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto add = make_shared<ov::op::v1::Add>(arg0, arg1);
+
+    return make_shared<ov::Function>(add, ov::ParameterVector{arg0, arg1});
+}
+
+std::shared_ptr<ov::Function> create_mul_model(ov::Shape shape) {
+    auto arg0 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto arg1 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto mul = make_shared<ov::op::v1::Multiply>(arg0, arg1);
+
+    return make_shared<ov::Function>(mul, ov::ParameterVector{arg0, arg1});
+}
+
+TEST(AsyncInferQueueOVTests, BasicFlowTest) {
+    ov::runtime::Core core;
+    ov::runtime::ExecutableNetwork net =
+        core.compile_model(create_add_model({8}), "CPU");
+
+    ov::runtime::AsyncInferQueue basic_queue(net);
+
+    size_t counter = 0;
+    std::vector<ov::Any> callback_data;
+
+    auto f = [&counter, &callback_data](std::exception_ptr e,
+                                        ov::runtime::InferRequest &request,
+                                        const ov::Any &data) {
+        try {
+            if (e) {
+                std::rethrow_exception(e);
+            }
+        } catch (const std::exception& err) {
+            throw ov::Exception(err.what());
+        }
+        counter += 1;
+        callback_data.push_back(data);
```

Comment:
Good catch, I will add mutex here from outside of lamba or just create a test where sufficient vector will be created. Which option do you prefer in that example?

Reply:
You can create a class with a mutex and a vector inside. Expose a "setter" to add results from a finished job (and lock a mutex in this setter). From the test perspective it will look like an atomic operation and you can reuse this in the rest of the tests where a similar approach is needed(you have quite a lot of them).